src/components/ha-icon-picker.ts

```diff
@@ -87,46 +131,37 @@ export class HaIconPicker extends LitElement {
     `;
   }
 
-  private async _openedChanged(ev: PolymerChangedEvent<boolean>) {
-    this._opened = ev.detail.value;
-    if (this._opened && !iconLoaded) {
-      const iconList = await import("../../build/mdi/iconList.json");
-
-      iconItems = iconList.default.map((icon) => ({
-        icon: `mdi:${icon.name}`,
-        keywords: icon.keywords,
-      }));
-      iconLoaded = true;
-
-      this.comboBox.filteredItems = iconItems;
-
-      Object.keys(customIcons).forEach((iconSet) => {
-        this._loadCustomIconItems(iconSet);
-      });
-    }
-  }
-
-  private async _loadCustomIconItems(iconsetPrefix: string) {
-    try {
-      const getIconList = customIcons[iconsetPrefix].getIconList;
-      if (typeof getIconList !== "function") {
-        return;
+  private _filterIcons = memoizeOne(
+    (filterString: string, iconItems = ICONS) => {
+      if (!filterString) {
+        return iconItems;
+      }
+      const startTime = performance.now();
+      const filteredItems = iconItems.filter(
+        (item) =>
+          item.icon.includes(filterString) ||
+          item.keywords.some((word) => word.includes(filterString))
+      );
```

Comment:
We didn't use filter because we wanted to display the matches by name before the matches by keyword.
If you type `nature` you prefer to have `mdi:nature` instead of `mdi:bee`

Reply:
My primary goal was to get rid of the duplicates and have one loop instead of two for performance.  For example, when you type "m" it searches 7096 icons and finds over 10k results.

As for ranking the includes over the keywords, I think that's flawed logic, especially when searching for words that are commonly part of other words.  For example, if I search for "rain", the icons I probably care about are buried below a whole bunch of "train" icons.  Or if I search for "hot", there are a bunch of screenshot icons to contend with before getting to the relevant ones.

I think trying to rank and sort them is a good idea, but the algorithm needs to be more nuanced and it should use a sorting function after the filter for good performance.

Reply:
I implemented a simple ranking when filtering.  This restores your preference to rank icon name matches over keywords, but also addresses the issues I pointed out by ranking exact matches higher than partial word inclusions.  Test it out and let me know what you think.